v2/src/main/scala/coop/rchain/v2/casper/data/FinalizationFringe.scala

```diff
@@ -0,0 +1,17 @@
+package coop.rchain.v2.casper.data
+
+/**
+ * Highest finalized messages for senders.
+ *
+ * Unlike [[LatestMessages]], there can be only one finalized message per sender in a fringe,
+ * otherwise it would mean finalizing equivocating messages.
+ */
+final case class FinalizationFringe[M, S](v: Map[S, M])
+
+object FinalizationFringe {
+
+  /**
+   * Find messages that are finalized in next compared to prev.
+   */
+  def diff[M, S](next: FinalizationFringe[M, S], prev: FinalizationFringe[M, S]): Set[M] = ???
+}
```

Comment:
Sorry that I am still kind of unkown on the term `FinalizedFringe`. Is it possible to explain it in an example?

Reply:
`Finalization fringe` is a set of messages: highest finalized message for each known sender. Known senders are defined by latest messages, which you use to run Casper.

<img width="1444" alt="Screenshot 2021-11-02 at 12 33 08" src="https://user-images.githubusercontent.com/1746367/139821639-1315163c-5850-45fb-9b20-24f66062f66f.png">

Here you can see grey blocks: they are finalized. Collectively they are finalization fringe. Everything below them is finalized (conflict resolution has been done and acceptances and rejects are final).
 
(Also you can see there on extra grey block at the bottom - this is highest common message for messages of the fringe, it is required to merge fringe and get finalized state).